File: src/main/java/org/web3j/protocol/aion/JsonRpc2_0Aion.java
package org.web3j.protocol.aion;

import java.math.BigInteger;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.ScheduledExecutorService;

import org.web3j.protocol.Web3jService;
import org.web3j.protocol.core.DefaultBlockParameter;
import org.web3j.protocol.core.DefaultBlockParameterName;
import org.web3j.protocol.core.JsonRpc2_0Web3j;
import org.web3j.protocol.core.Request;
import org.web3j.protocol.core.methods.request.ShhFilter;
import org.web3j.protocol.core.methods.request.Transaction;
import org.web3j.protocol.core.methods.response.DbGetHex;
import org.web3j.protocol.core.methods.response.DbGetString;
import org.web3j.protocol.core.methods.response.DbPutHex;
import org.web3j.protocol.core.methods.response.DbPutString;
import org.web3j.protocol.core.methods.response.EthBlock;
import org.web3j.protocol.core.methods.response.EthCompileLLL;
import org.web3j.protocol.core.methods.response.EthCompileSerpent;
import org.web3j.protocol.core.methods.response.EthGetBalance;
import org.web3j.protocol.core.methods.response.EthGetCode;
import org.web3j.protocol.core.methods.response.EthGetStorageAt;
import org.web3j.protocol.core.methods.response.EthGetTransactionCount;
import org.web3j.protocol.core.methods.response.EthGetUncleCountByBlockHash;
import org.web3j.protocol.core.methods.response.EthGetUncleCountByBlockNumber;
import org.web3j.protocol.core.methods.response.EthGetWork;
import org.web3j.protocol.core.methods.response.EthSubmitWork;
import org.web3j.protocol.core.methods.response.ShhAddToGroup;
import org.web3j.protocol.core.methods.response.ShhHasIdentity;
import org.web3j.protocol.core.methods.response.ShhMessages;
import org.web3j.protocol.core.methods.response.ShhNewFilter;
import org.web3j.protocol.core.methods.response.ShhNewGroup;
import org.web3j.protocol.core.methods.response.ShhNewIdentity;
import org.web3j.protocol.core.methods.response.ShhPost;
import org.web3j.protocol.core.methods.response.ShhUninstallFilter;
import org.web3j.protocol.core.methods.response.ShhVersion;
import org.web3j.utils.Numeric;

public class JsonRpc2_0Aion extends JsonRpc2_0Web3j implements Aion {

    public JsonRpc2_0Aion(final Web3jService web3jService) {
        super(web3jService);
    }

    public JsonRpc2_0Aion(
            final Web3jService web3jService,
            final long pollingInterval,
            final ScheduledExecutorService scheduledExecutorService) {
        super(web3jService, pollingInterval, scheduledExecutorService);
    }

    // Unsupported Endpoints

    @Override
    public Request<?, EthGetUncleCountByBlockHash> ethGetUncleCountByBlockHash(
            final String blockHash) {
        throw new UnsupportedOperationException("Spec-defined functions that deal with Uncles are not part of the Aion protocol.");
    }

    @Override
    public Request<?, EthGetUncleCountByBlockNumber> ethGetUncleCountByBlockNumber(
            final DefaultBlockParameter defaultBlockParameter) {
        throw new UnsupportedOperationException("Spec-defined functions that deal with Uncles are not part of the Aion protocol.");
    }

    @Override
    public Request<?, EthBlock> ethGetUncleByBlockHashAndIndex(
            final String blockHash, final BigInteger transactionIndex) {
        throw new UnsupportedOperationException("Spec-defined functions that deal with Uncles are not part of the Aion protocol.");
    }

    @Override
    public Request<?, EthBlock> ethGetUncleByBlockNumberAndIndex(
            final DefaultBlockParameter defaultBlockParameter, final BigInteger uncleIndex) {
        throw new UnsupportedOperationException("Spec-defined functions that deal with Uncles are not part of the Aion protocol.");
    }

    @Override
    public Request<?, EthCompileLLL> ethCompileLLL(
            final String sourceCode) {
        throw new UnsupportedOperationException("The LLL compiler is not supported. " +
                "Currently, the only language supported on Aion is Solidity-on-Aion-FVM.");
    }

    @Override
    public Request<?, EthCompileSerpent> ethCompileSerpent(
            final String sourceCode) {
        throw new UnsupportedOperationException("The Serpent compiler is not supported. " +
                "Currently, the only language supported on Aion is Solidity-on-Aion-FVM.");
    }

    @Override
    public Request<?, EthGetWork> ethGetWork() {
        throw new UnsupportedOperationException("Implemented as part of the stratum API submodule.");
    }

    @Override
    public Request<?, EthSubmitWork> ethSubmitWork(
            final String nonce, final String headerPowHash, final String mixDigest) {
        throw new UnsupportedOperationException("Implemented as part of the stratum API submodule.");
    }

    @Override
    public Request<?, ShhPost> shhPost(
            final org.web3j.protocol.core.methods.request.ShhPost shhPost) {
        throw new UnsupportedOperationException("Aion currently does not support the Whisper protocol.");
    }

    @Override
    public Request<?, ShhVersion> shhVersion() {
        throw new UnsupportedOperationException("Aion currently does not support the Whisper protocol.");
    }

    @Override
    public Request<?, ShhNewIdentity> shhNewIdentity() {
        throw new UnsupportedOperationException("Aion currently does not support the Whisper protocol.");
    }

    @Override
    public Request<?, ShhHasIdentity> shhHasIdentity(
            final String identityAddress) {
        throw new UnsupportedOperationException("Aion currently does not support the Whisper protocol.");
    }

    @Override
    public Request<?, ShhNewGroup> shhNewGroup() {
        throw new UnsupportedOperationException("Aion currently does not support the Whisper protocol.");
    }

    @Override
    public Request<?, ShhAddToGroup> shhAddToGroup(
            final String identityAddress) {
        throw new UnsupportedOperationException("Aion currently does not support the Whisper protocol.");
    }

    @Override
    public Request<?, ShhNewFilter> shhNewFilter(
            final ShhFilter shhFilter) {
        throw new UnsupportedOperationException("Aion currently does not support the Whisper protocol.");
    }

    @Override
    public Request<?, ShhUninstallFilter> shhUninstallFilter(
            final BigInteger filterId) {
        throw new UnsupportedOperationException("Aion currently does not support the Whisper protocol.");
    }

    @Override
    public Request<?, ShhMessages> shhGetFilterChanges(
            final BigInteger filterId) {
        throw new UnsupportedOperationException("Aion currently does not support the Whisper protocol.");
    }

    @Override
    public Request<?, ShhMessages> shhGetMessages(
            final BigInteger filterId) {
        throw new UnsupportedOperationException("Aion currently does not support the Whisper protocol.");
    }

    @Override
    public Request<?, DbPutString> dbPutString(
            final String databaseName, final String keyName, final String stringToStore) {
        throw new UnsupportedOperationException("Aion currently does not support the Whisper protocol.");
    }

    @Override
    @Deprecated
    public Request<?, DbGetString> dbGetString(
            final String databaseName, final String keyName) {
        throw new UnsupportedOperationException("The database APIs have been deprecated in the Ethereum Spec.");
    }

    @Override
    @Deprecated
    public Request<?, DbPutHex> dbPutHex(
            final String databaseName, final String keyName, final String dataToStore) {
        throw new UnsupportedOperationException("The database APIs have been deprecated in the Ethereum Spec.");
    }

    @Override
    @Deprecated
    public Request<?, DbGetHex> dbGetHex(
            final String databaseName, final String keyName) {
        throw new UnsupportedOperationException("The database APIs have been deprecated in the Ethereum Spec.");
    }

    // Implementation Deviations

    /**
     * @param defaultBlockParameter <code>pending</code> status is not supported for as a default block parameter.
     */
    @Override
    public Request<?, EthGetBalance> ethGetBalance(
            final String address, final DefaultBlockParameter defaultBlockParameter) {
        return new Request<>(
                "eth_getBalance",
                removePendingStatusParameter(defaultBlockParameter, address),
                web3jService,
                EthGetBalance.class);
    }

    /**
     * @param defaultBlockParameter <code>pending</code> status is not supported for as a default block parameter.
     */
    @Override
    public Request<?, EthGetCode> ethGetCode(
            final String address, final DefaultBlockParameter defaultBlockParameter) {
        return new Request<>(
                "eth_getCode",
                removePendingStatusParameter(defaultBlockParameter, address),
                web3jService,
                EthGetCode.class);
    }

    /**
     * @param defaultBlockParameter <code>pending</code> status is not supported for as a default block parameter.
     */
    @Override
    public Request<?, EthGetTransactionCount> ethGetTransactionCount(
            final String address, final DefaultBlockParameter defaultBlockParameter) {
        return new Request<>(
                "eth_getTransactionCount",
                removePendingStatusParameter(defaultBlockParameter, address),
                web3jService,
                EthGetTransactionCount.class);
    }

    /**
     * @param defaultBlockParameter <code>pending</code> status is not supported for as a default block parameter.
     */
    @Override
    public Request<?, EthGetStorageAt> ethGetStorageAt(
            String address, BigInteger position, DefaultBlockParameter defaultBlockParameter) {
        return new Request<>(
                "eth_getStorageAt",
                removePendingStatusParameter(defaultBlockParameter, address, Numeric.encodeQuantity(position)),
                web3jService,
                EthGetStorageAt.class);
    }

    /**
     * @param defaultBlockParameter <code>pending</code> status is not supported for as a default block parameter.
     */
    @Override
    public Request<?, org.web3j.protocol.core.methods.response.EthCall> ethCall(
            Transaction transaction, DefaultBlockParameter defaultBlockParameter) {
        return new Request<>(
                "eth_call",
                removePendingStatusParameter(defaultBlockParameter, transaction),
                web3jService,
                org.web3j.protocol.core.methods.response.EthCall.class);
    }

    private static List<?> removePendingStatusParameter(
            final DefaultBlockParameter defaultBlockParameter, final Object... arguments) {

        final List<Object> result = Arrays.asList(arguments);

        if (!DefaultBlockParameterName.PENDING.getValue()
                .equalsIgnoreCase(defaultBlockParameter.getValue())) {

            result.add(defaultBlockParameter.getValue());
        }

        return result;
    }
}

File: src/main/kotlin/org/web3j/protocol/aion/JsonRpc2_0Aion.kt
package org.web3j.protocol.aion

import mu.KLogging
import org.web3j.protocol.ObjectMapperFactory
import org.web3j.protocol.Web3jService
import org.web3j.protocol.core.DefaultBlockParameter
import org.web3j.protocol.core.DefaultBlockParameterName
import org.web3j.protocol.core.JsonRpc2_0Web3j
import org.web3j.protocol.core.Request
import org.web3j.protocol.core.methods.request.Transaction
import org.web3j.protocol.core.methods.response.EthCall
import org.web3j.protocol.core.methods.response.EthGetBalance
import org.web3j.protocol.core.methods.response.EthGetCode
import org.web3j.protocol.core.methods.response.EthGetStorageAt
import org.web3j.protocol.core.methods.response.EthGetTransactionCount
import org.web3j.utils.Numeric
import java.math.BigInteger
import java.util.Arrays
import java.util.concurrent.ScheduledExecutorService

@Suppress("ClassName")
internal class JsonRpc2_0Aion : JsonRpc2_0Web3j, Aion {

    init {
        ObjectMapperFactory.getObjectMapper()
            .addMixIn(Request::class.java, AionRequestMixIn::class.java)
    }

    constructor(web3jService: Web3jService) : super(web3jService)

    constructor(
        web3jService: Web3jService,
        pollingInterval: Long,
        scheduledExecutorService: ScheduledExecutorService
    ) : super(web3jService, pollingInterval, scheduledExecutorService)

    // Unsupported Endpoints
/*
    override fun ethGetUncleCountByBlockHash(blockHash: String)
            : Request<*, EthGetUncleCountByBlockHash> {
        throw UnsupportedOperationException("Spec-defined functions that deal with Uncles are not part of the Aion protocol.")
    }

    override fun ethGetUncleCountByBlockNumber(defaultBlockParameter: DefaultBlockParameter)
            : Request<*, EthGetUncleCountByBlockNumber> {
        throw UnsupportedOperationException("Spec-defined functions that deal with Uncles are not part of the Aion protocol.")
    }

    override fun ethGetUncleByBlockHashAndIndex(blockHash: String, transactionIndex: BigInteger)
            : Request<*, EthBlock> {
        throw UnsupportedOperationException("Spec-defined functions that deal with Uncles are not part of the Aion protocol.")
    }

    override fun ethGetUncleByBlockNumberAndIndex(
            defaultBlockParameter: DefaultBlockParameter,
            uncleIndex: BigInteger): Request<*, EthBlock> {
        throw UnsupportedOperationException("Spec-defined functions that deal with Uncles are not part of the Aion protocol.")
    }

    override fun ethCompileLLL(sourceCode: String): Request<*, EthCompileLLL> {
        throw UnsupportedOperationException("The LLL compiler is not supported. "
                + "Currently, the only language supported on Aion is Solidity-on-Aion-FVM.")
    }

    override fun ethCompileSerpent(sourceCode: String): Request<*, EthCompileSerpent> {
        throw UnsupportedOperationException("The Serpent compiler is not supported. "
                + "Currently, the only language supported on Aion is Solidity-on-Aion-FVM.")
    }

    override fun ethGetWork(): Request<*, EthGetWork> {
        throw UnsupportedOperationException("Implemented as part of the stratum API submodule.")
    }

    override fun ethSubmitWork(nonce: String, headerPowHash: String, mixDigest: String): Request<*, EthSubmitWork> {
        throw UnsupportedOperationException("Implemented as part of the stratum API submodule.")
    }

    override fun shhPost(shhPost: org.web3j.protocol.core.methods.request.ShhPost): Request<*, ShhPost> {
        throw UnsupportedOperationException("Aion currently does not support the Whisper protocol.")
    }

    override fun shhVersion(): Request<*, ShhVersion> {
        throw UnsupportedOperationException("Aion currently does not support the Whisper protocol.")
    }

    override fun shhNewIdentity(): Request<*, ShhNewIdentity> {
        throw UnsupportedOperationException("Aion currently does not support the Whisper protocol.")
    }

    override fun shhHasIdentity(identityAddress: String): Request<*, ShhHasIdentity> {
        throw UnsupportedOperationException("Aion currently does not support the Whisper protocol.")
    }

    override fun shhNewGroup(): Request<*, ShhNewGroup> {
        throw UnsupportedOperationException("Aion currently does not support the Whisper protocol.")
    }

    override fun shhAddToGroup(identityAddress: String): Request<*, ShhAddToGroup> {
        throw UnsupportedOperationException("Aion currently does not support the Whisper protocol.")
    }

    override fun shhNewFilter(shhFilter: ShhFilter): Request<*, ShhNewFilter> {
        throw UnsupportedOperationException("Aion currently does not support the Whisper protocol.")
    }

    override fun shhUninstallFilter(filterId: BigInteger): Request<*, ShhUninstallFilter> {
        throw UnsupportedOperationException("Aion currently does not support the Whisper protocol.")
    }

    override fun shhGetFilterChanges(filterId: BigInteger): Request<*, ShhMessages> {
        throw UnsupportedOperationException("Aion currently does not support the Whisper protocol.")
    }

    override fun shhGetMessages(filterId: BigInteger): Request<*, ShhMessages> {
        throw UnsupportedOperationException("Aion currently does not support the Whisper protocol.")
    }

    override fun dbPutString(databaseName: String, keyName: String, stringToStore: String): Request<*, DbPutString> {
        throw UnsupportedOperationException("Aion currently does not support the Whisper protocol.")
    }

    @Deprecated("")
    override fun dbGetString(databaseName: String, keyName: String): Request<*, DbGetString> {
        throw UnsupportedOperationException("The database APIs have been deprecated in the Ethereum Spec.")
    }

    @Deprecated("")
    override fun dbPutHex(databaseName: String, keyName: String, dataToStore: String): Request<*, DbPutHex> {
        throw UnsupportedOperationException("The database APIs have been deprecated in the Ethereum Spec.")
    }

    @Deprecated("")
    override fun dbGetHex(databaseName: String, keyName: String): Request<*, DbGetHex> {
        throw UnsupportedOperationException("The database APIs have been deprecated in the Ethereum Spec.")
    }*/

    // Implementation Deviations

    /**
     * @param defaultBlockParameter `pending` status is not supported for as a default block parameter.
     */
    override fun ethGetBalance(
        address: String,
        defaultBlockParameter: DefaultBlockParameter
    ):
        Request<*, EthGetBalance> {

        return Request<Any, EthGetBalance>(
            "eth_getBalance",
            removePendingStatusParameter(defaultBlockParameter, address),
            web3jService,
            EthGetBalance::class.java
        )
    }

    /**
     * @param defaultBlockParameter `pending` status is not supported for as a default block parameter.
     */
    override fun ethGetCode(
        address: String,
        defaultBlockParameter: DefaultBlockParameter
    ):
        Request<*, EthGetCode> {

        return Request<Any, EthGetCode>(
            "eth_getCode",
            removePendingStatusParameter(defaultBlockParameter, address),
            web3jService,
            EthGetCode::class.java
        )
    }

    /**
     * @param defaultBlockParameter `pending` status is not supported for as a default block parameter.
     */
    override fun ethGetTransactionCount(
        address: String,
        defaultBlockParameter: DefaultBlockParameter
    ):
        Request<*, EthGetTransactionCount> {

        return Request<Any, EthGetTransactionCount>(
            "eth_getTransactionCount",
            removePendingStatusParameter(defaultBlockParameter, address),
            web3jService,
            EthGetTransactionCount::class.java
        )
    }

    /**
     * @param defaultBlockParameter `pending` status is not supported for as a default block parameter.
     */
    override fun ethGetStorageAt(
        address: String,
        position: BigInteger,
        defaultBlockParameter: DefaultBlockParameter
    ):
        Request<*, EthGetStorageAt> {

        return Request<Any, EthGetStorageAt>(
            "eth_getStorageAt",
            removePendingStatusParameter(defaultBlockParameter, address, Numeric.encodeQuantity(position)),
            web3jService,
            EthGetStorageAt::class.java
        )
    }

    /**
     * @param defaultBlockParameter `pending` status is not supported for as a default block parameter.
     */
    override fun ethCall(
        transaction: Transaction,
        defaultBlockParameter: DefaultBlockParameter
    ):
        Request<*, EthCall> {

        return Request<Any, EthCall>(
            "eth_call",
            removePendingStatusParameter(defaultBlockParameter, transaction),
            web3jService,
            org.web3j.protocol.core.methods.response.EthCall::class.java
        )
    }

    private fun removePendingStatusParameter(
        defaultBlockParameter: DefaultBlockParameter,
        vararg arguments: Any
    ):
        List<*> {

        val result = Arrays.asList(*arguments)

        if (!DefaultBlockParameterName.PENDING.value
                .equals(defaultBlockParameter.value, ignoreCase = true)
        ) {

            result.add(defaultBlockParameter.value)
        }

        return result
    }

    companion object : KLogging()
}
